File: fibonacci.java
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Stream;
import java.util.stream.Collectors;

class Fibonacci {
  public static class FibonacciBeforeJava8 {
    public List<Integer> fibonacciSequence(int limit) {
      List<Integer> sequence = new ArrayList<>();
      int a = 0, b = 1;

      while (sequence.size() < limit) {
        sequence.add(a);
        int temp = a;
        a = b;
        b = temp + b;
      }

      return sequence;
    }

    public int fibonacciValueAtIndex(int index) {
      if (index <= 1) {
        return index;
      }

      int a = 0, b = 1;

      for (int i = 2; i <= index; i++) {
        int temp = a;
        a = b;
        b = temp + b;
      }

      return b;
    }

    public boolean isFibonacciNumber(int number) {
      if (number < 2) {
        return false;
      }
      int a = 0, b = 1;

      while (b < number) {
        int temp = a;
        a = b;
        b = temp + b;
      }

      return b == number;

    }

    public int fibonacciValueAtIndexWithRecursion(int index) {
      if (index <= 1) {
        return index;
      }

      return fibonacciValueAtIndexWithRecursion(index - 1) + fibonacciValueAtIndexWithRecursion(index - 2);
    }
  }

  public class FibonacciAfterJava8 {
    public List<Integer> fibonacciSequence(int limit) {
      return Stream.iterate(new int[] { 0, 1 }, f -> new int[] { f[1], f[0] + f[1] })
          .limit(limit)
          .map(f -> f[0])
          .collect(Collectors.toList());
    }

    public int fibonacciValutAtIndex(int index) {
      return Stream.iterate(new int[] { 0, 1 }, f -> new int[] { f[1], f[0] + f[1] })
          .limit(index + 1)
          .skip(index)
          .findFirst()
          .get()[0];
    }

    public boolean isFibonacciNumber(int number) {
      return Stream.iterate(new int[] { 0, 1 }, f -> new int[] { f[1], f[0] + f[1] })
          .map(f -> f[0])
          .anyMatch(f -> f == number);
    }
  }

  public static void main(String[] args) {
  }

}

File: fibonacci.py

"""
Fibonacci

Fibonacci Formula: Fn = Fn - 1 + Fn + -2

"""


def fibonacci_sequence(n):
    """
    Given a n value return fibonacci sequence until value index
    First two values it's 0 and 1
    Example of fibonacci sequence:
    0,1,1,2,3,5,8,13,21
    """
    sequence = [0, 1]
    while len(sequence) <= n:
        sequence.append(sequence[-1] + sequence[-2])
    return sequence[:n]


print(fibonacci_sequence(0))
print(fibonacci_sequence(1))
print(fibonacci_sequence(2))
print(fibonacci_sequence(3))
print(fibonacci_sequence(5))


def fibonacci_value_at_index(index):
    """
    Return value from fibonacci sequence from given index 
    """
    if index <= 1:
        return index
    a, b = 0, 1

    for _ in range(2, index+1):
        a, b = b, a + b
    return b


print(fibonacci_value_at_index(0))
print(fibonacci_value_at_index(1))
print(fibonacci_value_at_index(2))
print(fibonacci_value_at_index(3))
print(fibonacci_value_at_index(5))


def is_fibonacci_number(number):
    """
    Evaluate if a given number belongs to fibonacci sequence
    """
    if number < 0:
        return False
    a, b = 0, 1
    while b < number:
        a, b = b, a+b
    return b == number


print(is_fibonacci_number(0))
print(is_fibonacci_number(1))
print(is_fibonacci_number(5))
print(is_fibonacci_number(10))
print(is_fibonacci_number(23))
